mypyc/irbuild/expression.py

```diff
@@ -352,8 +352,32 @@ def transform_conditional_expr(builder: IRBuilder, expr: ConditionalExpr) -> Val
 
 
 def transform_comparison_expr(builder: IRBuilder, e: ComparisonExpr) -> Value:
-    # TODO: Don't produce an expression when used in conditional context
+    # x in (...)/[...]
+    if e.operators[0] in ['in', 'not in'] and isinstance(e.operands[1], (TupleExpr, ListExpr)):
+        cmp_op = '==' if e.operators[0] == 'in' else '!='
+        items = e.operands[1].items
+        n_items = len(items)
+        # x in []/() -> False
+        if n_items == 0:
+            name = 'builtins.False' if e.operators[0] == 'in' else 'builtins.True'
+            return builder.add(PrimitiveOp([], name_ref_ops[name], e.line))
+        # # x in [y]/(y) -> x == y
+        # # x not in [y]/(y) -> x != y
+        elif n_items == 1:
+            e.operators = [cmp_op]
+            e.operands[1] = items[0]
+        # x in y -> x == y[0] or ... or x == y[n]
+        # 16 is arbitrarily chosen to limit code size
+        elif n_items < 16:
+            bin_op = 'or' if e.operators[0] == 'in' else 'and'
+            lhs = e.operands[0]
+            exprs = (ComparisonExpr([cmp_op], [lhs, item]) for item in items)
```

Comment:
I /think/ that since these expressions don't appear in the type table, they will get coerced to object, leading to some kind of pointless boxing. I guess that isn't actually that expensive for bools, but it's worth cleaning up. Fine to just put a TODO in for now if you don't feel like cleaning it up now.

Reply:
Thanks for the review @msullivan, I'd be happy to fix this but I would probably need some more specific pointers of what needs to be modified for that to work.

Reply:
For what it's worth, `x in [1, 2, 3]` generates:

```
def f(i):
    i :: int
    r0, r1 :: object
    r2 :: short_int
    r3 :: bool
    r4 :: object
    r5 :: bool
    r6 :: short_int
    r7 :: bool
    r8 :: object
    r9 :: bool
    r10 :: short_int
    r11 :: bool
    r12 :: object
    r13 :: bool
L0:
    r2 = 1
    r3 = i == r2 :: int
    r4 = box(bool, r3)
    r5 = bool r4 :: object
    if r5 goto L1 else goto L2 :: bool
L1:
    r1 = r4
    goto L3
L2:
    r6 = 2
    r7 = i == r6 :: int
    r8 = box(bool, r7)
    r1 = r8
L3:
    r9 = bool r1 :: object
    if r9 goto L4 else goto L5 :: bool
L4:
    r0 = r1
    goto L6
L5:
    r10 = 3
    r11 = i == r10 :: int
    r12 = box(bool, r11)
    r0 = r12
L6:
    r13 = unbox(bool, r0)
    return r13
```

Reply:
@msullivan I ended up, perhaps somewhat hacky to shortcut all OpExpr/ComparisonExpr without types as bool_primitive. That removed the excessive box/unboxing, with these changes it's significantly faster, I measured somewhere between 46%-78% for micro benchmarks. Seems like it triggers some happy path in the C compiler, as the length of the tuple/list is no longer relevant for performance. (tested up to 16 items)

Reply:
Happy to hear it is a lot faster! I don't like this hack, though.

I think there are two ways forward:
 * Directly generate the code without creating an AST for it first. One way to do this would involve using `shortcircuit_helper`, though also it could probably just be done directly.
 * Put all of the generated expressions into the type table. This could probably be done ergonomically by adding a helper method that takes an expression and a type, adds it to the table, and returns the expression.

Historically though we haven't really done AST generation as part of compilation (I think mostly because it would require populating the type table), but it seems fine to allow if it is done tastefully.
(@JukkaL, do you agree?)

Reply:
No problem, I'll populate the type table with these types then, `shortcut_helper` scared me a bit, it seemed more straight forward to me to generate AST nodes for this issue.